src/aircraft_cmd.cpp

```diff
@@ -140,10 +140,24 @@ static StationID FindNearestHangar(const Aircraft *v)
 
 		/* v->tile can't be used here, when aircraft is flying v->tile is set to 0 */
 		uint distance = DistanceSquare(vtile, st->airport.tile);
-		if (v->acache.cached_max_range_sqr != 0) {
-			/* Check if our current destination can be reached from the depot airport. */
-			const Station *cur_dest = GetTargetAirportIfValid(v);
-			if (cur_dest != nullptr && DistanceSquare(st->airport.tile, cur_dest->airport.tile) > v->acache.cached_max_range_sqr) continue;
+
+		uint max_range = v->acache.cached_max_range_sqr;
+		if (max_range != 0) {
+			/* Determine destinations */
+			const Station *last_dest = GetTargetAirportIfValid(v);
+			const Station *next_dest;
+			if (v->current_order.IsType(OT_GOTO_STATION) ||
+					(v->current_order.IsType(OT_GOTO_DEPOT) && v->current_order.GetDepotActionType() != ODATFB_NEAREST_DEPOT)) {
+				next_dest = Station::GetIfValid(v->current_order.GetDestination());
+				if (next_dest == last_dest) last_dest = Station::GetIfValid(v->last_station_visited);
```

Comment:
The code might work, but the logic seems mightily twisted here.

We want to know where we came from and where we are going to, right? So why is the next destination put into `last_dest` and then queried again to be but into `next_dest`. And only if these two agree and it is in fact a goto order, the last station is actually determined.

If there's some sense in that I'd definitely want to see a comment shedding some light why `last_dest` is in fact often not the last destination.